app/launcher/Main.hs

```diff
@@ -0,0 +1,102 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE QuasiQuotes #-}
+
+module Main where
+
+import CLI
+    ( Decodable (decode), Encodable (encode), Network, Port (..), getArg )
+import Control.Concurrent
+    ( forkIO )
+import Control.Monad
+    ( forM, forM_, when )
+import GHC.IO.Handle
+    ( Handle )
+import Prelude
+import System.Console.Docopt
+    ( Docopt, docopt, exitWithUsage, isPresent, longOption, parseArgsOrExit )
+import System.Environment
+    ( getArgs )
+import System.Exit
+    ( exitWith )
+import System.Process
+    ( CreateProcess
+    , ProcessHandle
+    , cleanupProcess
+    , createProcess
+    , interruptProcessGroupOf
+    , proc
+    , waitForProcess
+    )
+
+
+-- | Command-Line Interface specification. See http://docopt.org/
+cli :: Docopt
+cli = [docopt|
+cardano-wallet-launcher
+
+Start the cardano wallet along with its API and underlying node.
+
+Requires cardano-http-bridge. To install, follow instructions at
+https://github.com/input-output-hk/cardano-http-bridge, and run
+  cargo install --path .
+in the directory.
+
+
+Usage:
+  cardano-wallet-launcher [options]
+  cardano-wallet-launcher --help
+
+Options:
+  --wallet-server-port <PORT>  port used for serving the wallet API [default: 8090]
+  --node-port <PORT>           port used for node-wallet communication [default: 8080]
+  --network <NETWORK>          mainnet or testnet [default: mainnet]
+|]
+
+
+type Process = (Maybe Handle, Maybe Handle, Maybe Handle, ProcessHandle)
+
+main :: IO ()
+main = do
+    args <- parseArgsOrExit cli =<< getArgs
+    when (args `isPresent` (longOption "help")) $ exitWithUsage cli
+    let getArg' = getArg args cli
+
+    nodePort <- getArg' (longOption "node-port") decode
+    walletPort <- getArg' (longOption "wallet-server-port") decode
+    network <- getArg' (longOption "network") decode
+
+    putStrLn $
+        "Starting wallet on port " ++ (encode walletPort) ++
+        ",\n         connecting to node on port " ++ (encode nodePort)
+
+    processes <- forM
+        [ nodeHttpBridgeOn nodePort network
+        , walletOn walletPort nodePort
+        ] createProcess
+
+    forM_ processes (forkIO . monitorAndKill processes)
+
+    monitorAndKill processes (head processes)
```

Comment:
I did this to get the main thread to wait, which is, I'll admit a bit sketchy.

Reply:
@KtorZ Could you have a look at the sanity of this `monitorAndKill` approach? 🙏 

I fork threads to run `waitForProcess` on each process, and kill everything if one exits.

Reply:
Looking at the PR now :eyes:

Reply:
This doesn't feel right. `monitorAndKill` waits for the first process to die and kill the others if something happens :thinking: ... What if the first process dies? The others will keep on running.
I'd advise to have a look at `Concurrently` and `mapConcurrently` from the `async` package; those are useful utils to do easy "supervision" of process or, at the very least, ensure that they are either all up, or all dead.

Reply:
A problem with `mapConcurrently` is that a process could exit successfully, without killing the others.

`monitorAndKill` is agnostic to this.

> What if the first process dies? The others will keep on running.

I'm actually not quite sure what you mean here. It contradicts 

> waits for the first process to die and kill the others if something happens

Previously, I ran `monitorAndKill` once more from the main thread waiting for `head processes` instead of just `forever $ threadDelay maxBound`. Might have been very confusing — sorry.

_will look into testing this regardless_